include/lsst/afw/typehandling/python.h

```diff
@@ -0,0 +1,65 @@
+/*
+ * This file is part of afw.
+ *
+ * Developed for the LSST Data Management System.
+ * This product includes software developed by the LSST Project
+ * (https://www.lsst.org).
+ * See the COPYRIGHT file at the top-level directory of this distribution
+ * for details of code ownership.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "pybind11/pybind11.h"
+
+#include "lsst/afw/typehandling/Storable.h"
+
+namespace lsst {
+namespace afw {
+namespace typehandling {
+
+/**
+ * "Trampoline" for Storable to let it be used as a base class in Python.
+ *
+ * Subclasses of Storable that are wrapped in %pybind11 should have a similar
+ * helper that subclasses `StorableHelper<subclass>`. This helper can be
+ * skipped if the subclass neither adds any virtual methods nor implements
+ * any abstract methods.
+ *
+ * @tparam Base the exact (most specific) class being wrapped
+ *
+ * @see [pybind11 documentation](https://pybind11.readthedocs.io/en/stable/advanced/classes.html)
+ */
+template <class Base = Storable>
+class StorableHelper : public Base {
+public:
+    // Can't wrap clone(); PYBIND11_OVERLOAD chokes on unique_ptr return type
```

Comment:
I'm not sure what you mean. How can `deepcopy` copy a C++ object if we don't implement `__deepcopy__` for it?

I'm pretty sure this is just a bug in `PYBIND11_OVERLOAD`, since the following works in the regular wrapper:

    cls.def("__copy__", &Storable::clone);

Reply:
I've been pondering how the pybind11 overload stuff could possibly work, given that that macro has got to somehow get at the `PyObject*` that holds `this` in order to call any Python-side implementation on the right instance.  The only idea I've come up with is that pybind11 actually further subclasses the helper with a derived class that holds that `PyObject*`, and the macro expands to something that downcasts `this` to that type (or if not inheritance, then something scarier, like C-style composition-based "inheritance" with pointer arithmetic).  That would explain why overriding something that returns a new instance is impossible, though that would be nice if it was documented.

Anyhow, I think that just amounts to this being too hard to try to deal with at all right now (if not impossible).